src/flow/src/repr.rs

```diff
@@ -0,0 +1,170 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! basically a wrapper around the `datatype` crate
+//! for basic Data Representation
+
+mod relation;
+
+use std::borrow::Borrow;
+use std::slice::SliceIndex;
+
+use api::helper::{pb_value_to_value_ref, value_to_grpc_value};
+use api::v1::Row as ProtoRow;
+use datatypes::data_type::ConcreteDataType;
+use datatypes::types::cast::CastOption;
+use datatypes::types::cast_with_opt;
+use datatypes::value::Value;
+use itertools::Itertools;
+pub(crate) use relation::{RelationDesc, RelationType};
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+
+use crate::expr::error::{CastValueSnafu, EvalError};
+
+/// System-wide Record count difference type.
+pub type Diff = i64;
+
+/// System-wide default timestamp type
+pub type Timestamp = i64;
+
+/// Default type for a repr of changes to a collection.
+pub type DiffRow = (Row, Timestamp, Diff);
+
+/// Convert a value that is or can be converted to Datetime to internal timestamp
+pub fn value_to_internal_ts(value: Value) -> Result<Timestamp, EvalError> {
+    match value {
+        Value::DateTime(ts) => Ok(ts.val()),
+        arg => {
+            let arg_ty = arg.data_type();
+            let res = cast_with_opt(
+                arg,
+                &ConcreteDataType::datetime_datatype(),
+                &CastOption { strict: true },
+            )
+            .context({
+                CastValueSnafu {
+                    from: arg_ty,
+                    to: ConcreteDataType::datetime_datatype(),
+                }
+            })?;
+            if let Value::DateTime(ts) = res {
+                Ok(ts.val())
+            } else {
+                unreachable!()
+            }
+        }
+    }
+}
+
+/// A row is a vector of values.
+///
+/// TODO(discord9): use a more efficient representation
+/// i.e. more compact like raw u8 of \[tag0, value0, tag1, value1, ...\]
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Default, Serialize, Deserialize)]
+pub struct Row {
+    pub inner: Vec<Value>,
+}
+
+impl Row {
+    pub fn empty() -> Self {
+        Self { inner: vec![] }
+    }
+    pub fn new(row: Vec<Value>) -> Self {
+        Self { inner: row }
+    }
+    pub fn get(&self, idx: usize) -> Option<&Value> {
+        self.inner.get(idx)
+    }
+    pub fn clear(&mut self) {
+        self.inner.clear();
+    }
+    /// clear and return the inner vector
+    ///
+    /// useful if you want to reuse the vector as a buffer
+    pub fn packer(&mut self) -> &mut Vec<Value> {
+        self.inner.clear();
+        &mut self.inner
+    }
+    /// pack a iterator of values into a row
+    pub fn pack<I>(iter: I) -> Row
+    where
+        I: IntoIterator<Item = Value>,
+    {
+        Self {
+            inner: iter.into_iter().collect(),
+        }
+    }
+    /// unpack a row into a vector of values
+    pub fn unpack(&self) -> Vec<Value> {
+        self.inner.clone()
+    }
```

Comment:
I wonder if it should reuse the `Row` after `unpack`. If not, I suggest taking the ownership of `Row`.

```suggestion
    /// unpack a row into a vector of values
    pub fn unpack(self) -> Vec<Value> {
        self.inner
    }
```